### PR TITLE
Feature/backoff

### PR DESCRIPTION
**Pull request by benjigoldberg**
Implements exponential backoff for the HTTP client on 5xx responses by default.

### CODE REVIEW

File: http/client.go [hunk 1]
```diff
@@ -30,12 +33,44 @@ type MiddlewareRoundTripper struct {
...
+	RetriableStatusCodes []int
+	InitialInterval      time.Duration
+	RandomizationFactor  float64
+	Multiplier           float64
+	MaxInterval          time.Duration
+	MaxRetries           uint64
+}
+
 // NewDefaultClient constructs the default HTTP Client with middleware. Providing an HTTP
 // RoundTripper is optional. If `nil` is received, the DefaultClient will be used.
```
Comment (briancharous) [inline http/client.go]: Document the default configuration for retries

File: http/client.go [hunk 2]
```diff
@@ -30,12 +33,44 @@ type MiddlewareRoundTripper struct {
...
+			http.StatusHTTPVersionNotSupported,
+			http.StatusVariantAlsoNegotiates,
+			http.StatusInsufficientStorage,
+			http.StatusLoopDetected,
+			http.StatusNotExtended,
+			http.StatusNetworkAuthenticationRequired,
+		},
+		InitialInterval:     100 * time.Millisecond,
+		Multiplier:          2,
+		MaxInterval:         30 * time.Second,
```
Comment (briancharous) [inline http/client.go]: This seems quite high, no?

File: http/client.go [hunk 3]
```diff
@@ -30,12 +33,44 @@ type MiddlewareRoundTripper struct {
...
 
+// RetryRoundTripper wraps a roundtripper with retry logic
+type RetryRoundTripper struct {
+	RoundTripper         http.RoundTripper
+	RetriableStatusCodes []int
+	InitialInterval      time.Duration
+	RandomizationFactor  float64
+	Multiplier           float64
+	MaxInterval          time.Duration
+	MaxRetries           uint64
```
Comment (briancharous) [inline http/client.go]: probably nitpicky but seems like `uint8` would be fine here. I can't imagine wanting to retry more than 2^8 times.

File: http/client.go [hunk 4]
```diff
@@ -30,12 +33,44 @@ type MiddlewareRoundTripper struct {
 	Middleware []ClientMiddleware
 }
 
+// RetryRoundTripper wraps a roundtripper with retry logic
+type RetryRoundTripper struct {
+	RoundTripper         http.RoundTripper
+	RetriableStatusCodes []int
```
Comment (briancharous) [inline http/client.go]: I guess this list is small, but should it be changed to a set?

File: http/client.go [hunk 5]
```diff
@@ -30,12 +33,44 @@ type MiddlewareRoundTripper struct {
...
+
 // NewDefaultClient constructs the default HTTP Client with middleware. Providing an HTTP
 // RoundTripper is optional. If `nil` is received, the DefaultClient will be used.
 func NewDefaultClient(metrics Metrics, roundTripper http.RoundTripper) http.Client {
 	if roundTripper == nil {
 		roundTripper = http.DefaultTransport
 	}
+	retryRoundTripper := RetryRoundTripper{
+		RoundTripper: roundTripper,
+		RetriableStatusCodes: []int{
```
Comment (briancharous) [inline http/client.go]: It seems like 505-511 (`StatusHTTPVersionNotSupported` - `StatusNetworkAuthenticationRequired`) aren't actually retry-able.

File: http/client.go (hunk 3 above)
Comment (benjigoldberg) [reply http/client.go]: I know, this is driven by the library wanting a uint64. I also laughed at this. I'll change it to uint8 and cast it internally.

File: http/client.go (hunk 4 above)
Comment (benjigoldberg) [reply http/client.go]: Oh man, how did I not think of that 🤦‍♂  -- great call out, I'll update

File: http/client.go (hunk 5 above)
Comment (benjigoldberg) [reply http/client.go]: Yep, good call.

File: http/client.go (hunk 2 above)
Comment (benjigoldberg) [reply http/client.go]: well, its the minimum threshold of either up to a 30 second backoff or 5 retries, whichever is reached first. I could be talked into a lower interval. The default in this library is 15 minutes!

Comment (benjigoldberg) [reply http/client.go]: Updated to 10 seconds

Comment (briancharous) [reply http/client.go]: Ah ok. Maybe 30s is fine then. I was unsure if 30s was the total amount of time or the amount of time for one retry.

Comment (benjigoldberg) [reply http/client.go]: I'm wrong, this is the minimum threshold for any individual backoff, NOT the total. Leaving at 10s.

File: http/client.go [hunk 6]
```diff
@@ -30,12 +33,45 @@ type MiddlewareRoundTripper struct {
...
+	MaxInterval          time.Duration
+	MaxRetries           uint8
+}
+
 // NewDefaultClient constructs the default HTTP Client with middleware. Providing an HTTP
 // RoundTripper is optional. If `nil` is received, the DefaultClient will be used.
+//
+// By default, the client is provides exponential backoff on 500-504 errors. The default
+// configuration for exponential backoff is to start with an interval of 100 milliseconds, a
+// multiplier of two, a randomization factor of 0.5 (for jitter), a max interval of 10 seconds,
```
Comment (wilsoniya) [inline http/client.go]: it's not clear to me from this description exactly how `RandomizationFactor` works.

File: http/client.go [hunk 7]
```diff
@@ -86,3 +122,55 @@ func (mrt MiddlewareRoundTripper) RoundTrip(req *http.Request) (*http.Response,
...
+
+	var resp *http.Response
+	var err error
+	makeRequestRetriable := func() error {
+		resp, err = rrt.RoundTripper.RoundTrip(req)
+
+		// If an error was encountered, retry. This typically indicates a failure to get a
+		// response.
+		if err != nil {
+			log.Get(req.Context()).Debug("retrying failed http request", zap.Error(err))
```
Comment (wilsoniya) [inline http/client.go]: This log message seems misleading because this implementation will only retry if the status code is in `rrt.RetriableStatusCodes`.

Comment (benjigoldberg) [reply http/client.go]: @wilsoniya this actually is a retry on a failed request error. We return an err so the retry occurs here as well

**Pull request merged**